[Bug][v0.2]: memmachine.semantic_memory.semantic_memory - background task keep crashing

### Describe the bug

[Bug][v0.2]: memmachine.semantic_memory.semantic_memory - background task keep crashing

### Steps to reproduce

**Step 1: add memory**
```
yufang@Memverges-MacBook-Pro MemMachine % curl -X POST "http://localhost:8080/api/v2/memories" \        
-H "Content-Type: application/json" \
-d '{
    "org_id": "org_sample",
    "project_id": "project_sample",
    "messages": [
        {
            "content": "This is to test memory content from user_sample to agent_sample",
            "producer": "user_sample",
            "produced_for": "",
            "timestamp": "2019-08-24T14:15:22Z",
            "metadata": {
                "department": "AI Research",
                "priority": "high",
                "agent_id": "agent_sample",
                "user_id": "user_sample"
            }
        }
    ]
}'
{"results":[{"uid":"124"}]}%  
```

**Step 2: delete project**
```
yufang@Memverges-MacBook-Pro MemMachine %  curl -X POST "http://localhost:8080/api/v2/projects/delete" \
-H "Content-Type: application/json" \
-d '{
  "org_id": "org_sample",  
  "project_id": "project_sample"   
}'               
yufang@Memverges-MacBook-Pro MemMachine
```

**Step 3: check postgresql database**
```
postgres=# select * from set_ingested_history;
                set_id                 | history_id | ingested 
---------------------------------------+------------+----------
 mem_session_org_sample/project_sample | 124        | f
(1 row)

postgres=# select * from episodestore;
 id | content | session_key | producer_id | producer_role | produced_for_id | episode_type | metadata | created_at 
----+---------+-------------+-------------+---------------+-----------------+--------------+----------+------------
(0 rows)

postgres=# 
```

**Step 4: add more memory**
```
yufang@Memverges-MacBook-Pro MemMachine % curl -X POST "http://localhost:8080/api/v2/memories" \
-H "Content-Type: application/json" \
-d '{
    "org_id": "org_sample",
    "project_id": "project_sample",
    "messages": [
        {
            "content": "This is to test memory content from user_sample to agent_sample",
            "producer": "user_sample"
        },
        {
            "content": "This is to test memory content from user_sample to agent_sample",
            "producer": "user_sample"
        },
        {
            "content": "This is to test memory content from user_sample to agent_sample",
            "producer": "user_sample"
        },
        {
            "content": "This is to test memory content from user_sample to agent_sample",
            "producer": "user_sample"
        }
    ]
}'
{"results":[{"uid":"125"},{"uid":"126"},{"uid":"127"},{"uid":"128"}]}%
```

**Step 5: app.py, keep crashing at memmachine.semantic_memory.semantic_memory**
```
2025-11-25 16:36:09,816 [ERROR] memmachine.semantic_memory.semantic_memory - background task crashed, restarting
  + Exception Group Traceback (most recent call last):
  |   File "/Users/yufang/test/MemMachine/src/memmachine/semantic_memory/semantic_memory.py", line 267, in _background_ingestion_task
  |     await ingestion_service.process_set_ids(dirty_sets)
  |   File "/Users/yufang/test/MemMachine/src/memmachine/semantic_memory/semantic_ingestion.py", line 66, in process_set_ids
  |     raise ExceptionGroup("Failed to process set ids", errors)
  | ExceptionGroup: Failed to process set ids (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/yufang/test/MemMachine/src/memmachine/semantic_memory/semantic_ingestion.py", line 91, in _process_single_set
    |     raise ValueError("Failed to retrieve messages. Invalid history_ids")
    | ValueError: Failed to retrieve messages. Invalid history_ids
    +------------------------------------
```

**Step 6: in postgresql**
```
postgres=# select * from set_ingested_history;
                set_id                 | history_id | ingested 
---------------------------------------+------------+----------
 mem_session_org_sample/project_sample | 125        | f
 mem_session_org_sample/project_sample | 126        | f
 mem_session_org_sample/project_sample | 127        | f
 mem_session_org_sample/project_sample | 128        | f
 mem_session_org_sample/project_sample | 124        | f
(5 rows)

postgres=# select * from episodestore;
 id  |                             content                             |        session_key        | producer_id | producer_role | produced_for_id | episode_type | metadata |          created_at           
-----+-----------------------------------------------------------------+---------------------------+-------------+---------------+-----------------+--------------+----------+-------------------------------
 125 | This is to test memory content from user_sample to agent_sample | org_sample/project_sample | user_sample |               |                 | MESSAGE      | {}       | 2025-11-25 16:24:35.024757-08
 126 | This is to test memory content from user_sample to agent_sample | org_sample/project_sample | user_sample |               |                 | MESSAGE      | {}       | 2025-11-25 16:24:35.024771-08
 127 | This is to test memory content from user_sample to agent_sample | org_sample/project_sample | user_sample |               |                 | MESSAGE      | {}       | 2025-11-25 16:24:35.024775-08
 128 | This is to test memory content from user_sample to agent_sample | org_sample/project_sample | user_sample |               |                 | MESSAGE      | {}       | 2025-11-25 16:24:35.024778-08
(4 rows)

postgres=# 

```

### Expected behavior

Should not crash

### Environment

MacOS: 15.6.1
MemMachine: c79799c [v0.2]

### Additional context

_No response_

--- a/src/memmachine/main/memmachine.py
+++ b/src/memmachine/main/memmachine.py
@@ -209,10 +209,12 @@ class MemMachine:
             semantic_memory_manager = (
                 await self._resources.get_semantic_session_manager()
             )
-
-            await semantic_memory_manager.delete_feature_set(
-                session_data=session_data,
-                memory_type=[IsolationType.SESSION],
+            await asyncio.gather(
+                semantic_memory_manager.delete_feature_set(
+                    session_data=session_data,
+                    memory_type=[IsolationType.SESSION],
+                ),
+                semantic_memory_manager.delete_messages(session_data=session_data),
             )
 
         tasks = [

--- a/src/memmachine/semantic_memory/semantic_memory.py
+++ b/src/memmachine/semantic_memory/semantic_memory.py
@@ -137,6 +137,9 @@ class SemanticService:
 
         _consolidate_errors_and_raise(res, "Failed to add message to sets")
 
+    async def delete_messages(self, *, set_ids: list[SetIdT]) -> None:
+        await self._semantic_storage.delete_history_set(set_ids=set_ids)
+
     async def number_of_uningested(self, set_ids: list[SetIdT]) -> int:
         return await self._semantic_storage.get_history_messages_count(
             set_ids=set_ids,

--- a/src/memmachine/semantic_memory/semantic_session_manager.py
+++ b/src/memmachine/semantic_memory/semantic_session_manager.py
@@ -147,6 +147,16 @@ class SemanticSessionManager:
 
         await asyncio.gather(*tasks)
 
+    async def delete_messages(
+        self,
+        session_data: SessionData,
+        *,
+        memory_type: list[IsolationType] = ALL_MEMORY_TYPES,
+    ) -> None:
+        set_ids = self._get_set_ids(session_data, memory_type)
+
+        await self._semantic_service.delete_messages(set_ids=set_ids)
+
     async def search(
         self,
         message: str,

--- a/src/memmachine/semantic_memory/storage/neo4j_semantic_storage.py
+++ b/src/memmachine/semantic_memory/storage/neo4j_semantic_storage.py
@@ -29,7 +29,7 @@ from memmachine.common.filter.filter_parser import (
 from memmachine.common.filter.filter_parser import (
     Or as FilterOr,
 )
-from memmachine.semantic_memory.semantic_model import SemanticFeature
+from memmachine.semantic_memory.semantic_model import SemanticFeature, SetIdT
 from memmachine.semantic_memory.storage.storage_base import (
     FeatureIdT,
     SemanticStorage,
@@ -577,6 +577,19 @@ class Neo4jSemanticStorage(SemanticStorage):
             DELETE h
             """,
             history_ids=[str(history_id) for history_id in history_ids],
+        )
+
+    async def delete_history_set(self, set_ids: list[SetIdT]) -> None:
+        if not set_ids:
+            return
+
+        await self._driver.execute_query(
+            """
+            MATCH (h:SetHistory)
+            WHERE h.set_id IN $set_ids
+            DELETE h
+            """,
+            set_ids=[str(set_id) for set_id in set_ids],
         )
 
     async def mark_messages_ingested(

--- a/src/memmachine/semantic_memory/storage/sqlalchemy_pgvector_semantic.py
+++ b/src/memmachine/semantic_memory/storage/sqlalchemy_pgvector_semantic.py
@@ -485,6 +485,18 @@ class SqlAlchemyPgVectorSemanticStorage(SemanticStorage):
             await session.execute(stmt)
             await session.commit()
 
+    async def delete_history_set(self, set_ids: list[SetIdT]) -> None:
+        if not set_ids:
+            return
+
+        stmt = delete(SetIngestedHistory).where(
+            SetIngestedHistory.set_id.in_(set_ids),
+        )
+
+        async with self._create_session() as session:
+            await session.execute(stmt)
+            await session.commit()
+
     def _apply_history_filter(
         self,
         stmt: Select,

--- a/src/memmachine/semantic_memory/storage/storage_base.py
+++ b/src/memmachine/semantic_memory/storage/storage_base.py
@@ -155,6 +155,10 @@ class SemanticStorage(ABC):
         raise NotImplementedError
 
     @abstractmethod
+    async def delete_history_set(self, set_ids: list[SetIdT]) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
     async def mark_messages_ingested(
         self,
         *,

--- a/tests/memmachine/main/test_memachine.py
+++ b/tests/memmachine/main/test_memachine.py
@@ -1,11 +1,13 @@
 from dataclasses import dataclass
 from datetime import UTC, datetime, timedelta
+from uuid import uuid4
 
 import numpy as np
 import pytest
 
 from memmachine import MemMachine
 from memmachine.common.episode_store import EpisodeEntry
+from memmachine.common.filter.filter_parser import parse_filter
 from memmachine.main.memmachine import MemoryType
 
 
@@ -142,3 +144,192 @@ async def test_memmachine_list_search_paginates_semantic(memmachine: MemMachine)
     finally:
         await semantic_storage.delete_features(feature_ids)
         await memmachine.delete_session(session_info)
+
+
+@pytest.mark.asyncio
+@pytest.mark.integration
+async def test_memmachine_create_get_and_delete_session(memmachine: MemMachine):
+    session_key = f"session-{uuid4()}"
+    delete_handle = _TempSession(
+        user_profile_id=session_key,
+        session_id=session_key,
+        role_profile_id=None,
+        session_key=session_key,
+    )
+    deleted = False
+
+    try:
+        session_info = await memmachine.create_session(
+            session_key,
+            description="integration-session",
+        )
+
+        assert session_info.description == "integration-session"
+        assert session_info.episode_memory_conf.session_key == session_key
+
+        fetched = await memmachine.get_session(session_key)
+        assert fetched is not None
+        assert fetched.description == "integration-session"
+
+        await memmachine.delete_session(delete_handle)
+        deleted = True
+        assert await memmachine.get_session(session_key) is None
+    finally:
+        if not deleted:
+            remaining = await memmachine.get_session(session_key)
+            if remaining is not None:
+                await memmachine.delete_session(delete_handle)
+
+
+@pytest.mark.asyncio
+@pytest.mark.integration
+async def test_memmachine_search_sessions_filters_metadata(memmachine: MemMachine):
+    session_manager = await memmachine._resources.get_session_data_manager()
+    created_sessions: list[str] = []
+
+    try:
+        for topic in ("alpha", "beta"):
+            new_session_key = f"metadata-session-{uuid4()}"
+            created_sessions.append(new_session_key)
+            await session_manager.create_new_session(
+                session_key=new_session_key,
+                configuration={"scope": "integration"},
+                param=memmachine._with_default_episodic_memory_conf(
+                    session_key=new_session_key
+                ),
+                description=f"session-{topic}",
+                metadata={"topic": topic},
+            )
+
+        all_sessions = await memmachine.search_sessions()
+        assert created_sessions[0] in all_sessions
+        assert created_sessions[1] in all_sessions
+
+        filter_expr = parse_filter("topic = 'alpha'")
+        filtered = await memmachine.search_sessions(search_filter=filter_expr)
+        assert set(filtered) == {created_sessions[0]}
+    finally:
+        for key in created_sessions:
+            cleanup_session = _TempSession(
+                user_profile_id=key,
+                session_id=key,
+                role_profile_id=None,
+                session_key=key,
+            )
+            remaining = await memmachine.get_session(key)
+            if remaining is not None:
+                await memmachine.delete_session(cleanup_session)
+
+
+@pytest.mark.asyncio
+@pytest.mark.integration
+async def test_memmachine_list_search_filters_metadata(
+    memmachine: MemMachine,
+    session_data,
+):
+    episode_ids = await memmachine.add_episodes(
+        session_data,
+        [
+            EpisodeEntry(
+                content="hello there",
+                producer_id="user",
+                producer_role="assistant",
+                metadata={"topic": "greeting"},
+            ),
+            EpisodeEntry(
+                content="status update",
+                producer_id="user",
+                producer_role="assistant",
+                metadata={"topic": "status"},
+            ),
+        ],
+        target_memories=[],
+    )
+
+    try:
+        filtered = await memmachine.list_search(
+            session_data=session_data,
+            target_memories=[MemoryType.Episodic],
+            search_filter="metadata.topic = 'greeting'",
+        )
+
+        assert filtered.episodic_memory is not None
+        assert [episode.content for episode in filtered.episodic_memory] == [
+            "hello there"
+        ]
+    finally:
+        await memmachine.delete_episodes(episode_ids, session_data=session_data)
+
+
+@pytest.mark.asyncio
+@pytest.mark.integration
+async def test_memmachine_delete_episodes_removes_history(
+    memmachine: MemMachine,
+    session_data,
+):
+    episode_ids = await memmachine.add_episodes(
+        session_data,
+        [
+            EpisodeEntry(
+                content="first",
+                producer_id="user",
+                producer_role="assistant",
+            ),
+            EpisodeEntry(
+                content="second",
+                producer_id="user",
+                producer_role="assistant",
+            ),
+        ],
+        target_memories=[],
+    )
+    deleted = False
+
+    try:
+        before_delete = await memmachine.list_search(
+            session_data=session_data,
+            target_memories=[MemoryType.Episodic],
+        )
+        assert before_delete.episodic_memory is not None
+        assert len(before_delete.episodic_memory) == 2
+
+        await memmachine.delete_episodes(episode_ids, session_data=session_data)
+        deleted = True
+
+        after_delete = await memmachine.list_search(
+            session_data=session_data,
+            target_memories=[MemoryType.Episodic],
+        )
+        assert after_delete.episodic_memory == []
+    finally:
+        if not deleted:
+            await memmachine.delete_episodes(episode_ids, session_data=session_data)
+
+
+@pytest.mark.asyncio
+@pytest.mark.integration
+async def test_memmachine_delete_features_removes_semantic_entries(
+    memmachine: MemMachine,
+    session_data,
+):
+    semantic_service = await memmachine._resources.get_semantic_service()
+    semantic_storage = semantic_service._semantic_storage
+    user_set_id = f"mem_user_{session_data.user_profile_id}"
+
+    feature_id = await semantic_storage.add_feature(
+        set_id=user_set_id,
+        category_name="profile",
+        feature="alias",
+        value="integration alias",
+        tag="facts",
+        embedding=np.array([0.5, 0.5], dtype=float),
+    )
+
+    try:
+        assert await semantic_storage.get_feature(feature_id) is not None
+        await memmachine.delete_features([feature_id])
+        assert await semantic_storage.get_feature(feature_id) is None
+    finally:
+        leftover = await semantic_storage.get_feature(feature_id)
+        if leftover is not None:
+            await memmachine.delete_features([feature_id])

--- a/tests/memmachine/main/test_memmachine_integration.py
+++ b/tests/memmachine/main/test_memmachine_integration.py
@@ -155,14 +155,10 @@ class TestMemMachineLongMemEval:
     @pytest.mark.integration
     async def test_memmachine_smoke_ingests_all_memories(
         self,
-        memmachine_config: Configuration,
+        memmachine: MemMachine,
         session_data,
         long_mem_conversations,
     ) -> None:
-        memmachine = MemMachine(memmachine_config)
-        await memmachine.start()
-        await memmachine.create_session(session_data.session_key)
-
         semantic_service = await memmachine._resources.get_semantic_service()
         semantic_service._feature_update_message_limit = 0
 
@@ -170,25 +166,22 @@ class TestMemMachineLongMemEval:
         if len(smoke_convo) > 2:
             smoke_convo = smoke_convo[:2]
 
-        try:
-            await self._ingest_conversations(
-                memmachine,
-                session_data,
-                [smoke_convo],
-            )
+        await self._ingest_conversations(
+            memmachine,
+            session_data,
+            [smoke_convo],
+        )
 
-            await self._wait_for_semantic_features(
-                memmachine, session_data, timeout_seconds=120
-            )
+        await self._wait_for_semantic_features(
+            memmachine, session_data, timeout_seconds=120
+        )
 
-            list_result = await memmachine.list_search(
-                session_data,
-                target_memories=[MemoryType.Semantic, MemoryType.Episodic],
-            )
+        list_result = await memmachine.list_search(
+            session_data,
+            target_memories=[MemoryType.Semantic, MemoryType.Episodic],
+        )
 
-            assert list_result.semantic_memory, "Semantic memory returned no features"
-            assert len(list_result.semantic_memory) > 0
-            assert list_result.episodic_memory is not None
-            assert len(list_result.episodic_memory) > 0
-        finally:
-            await memmachine.stop()
+        assert list_result.semantic_memory, "Semantic memory returned no features"
+        assert len(list_result.semantic_memory) > 0
+        assert list_result.episodic_memory is not None
+        assert len(list_result.episodic_memory) > 0

--- a/tests/memmachine/semantic_memory/storage/in_memory_semantic_storage.py
+++ b/tests/memmachine/semantic_memory/storage/in_memory_semantic_storage.py
@@ -421,6 +421,25 @@ class InMemorySemanticStorage(SemanticStorage):
                     if not history_map:
                         self._set_history_map.pop(set_id, None)
 
+    async def delete_history_set(self, set_ids: list[SetIdT]) -> None:
+        if not set_ids:
+            return
+
+        async with self._lock:
+            for set_id in set_ids:
+                history_map = self._set_history_map.pop(set_id, None)
+                if history_map is None:
+                    continue
+
+                for history_id in list(history_map.keys()):
+                    sets_map = self._history_to_sets.get(history_id)
+                    if sets_map is None:
+                        continue
+
+                    sets_map.pop(set_id, None)
+                    if not sets_map:
+                        self._history_to_sets.pop(history_id, None)
+
     async def mark_messages_ingested(
         self,
         *,

--- a/tests/memmachine/semantic_memory/storage/test_semantic_storage.py
+++ b/tests/memmachine/semantic_memory/storage/test_semantic_storage.py
@@ -495,6 +495,65 @@ async def test_delete_history_removes_set_associations(
 
 
 @pytest.mark.asyncio
+async def test_delete_history_set_removes_target_set(
+    semantic_storage: SemanticStorage,
+    episode_storage,
+):
+    h1_id = await _add_episode(episode_storage, content="first")
+    h2_id = await _add_episode(episode_storage, content="second")
+
+    await semantic_storage.add_history_to_set(set_id="alpha", history_id=h1_id)
+    await semantic_storage.add_history_to_set(set_id="alpha", history_id=h2_id)
+    await semantic_storage.add_history_to_set(set_id="beta", history_id=h1_id)
+    await semantic_storage.add_history_to_set(set_id="beta", history_id=h2_id)
+
+    assert set(await semantic_storage.get_history_set_ids()) == {"alpha", "beta"}
+
+    await semantic_storage.delete_history_set(set_ids=["alpha"])
+
+    assert await semantic_storage.get_history_messages(set_ids=["alpha"]) == []
+    remaining = await semantic_storage.get_history_messages(set_ids=["beta"])
+    assert set(map(str, remaining)) == {str(h1_id), str(h2_id)}
+    assert set(await semantic_storage.get_history_set_ids()) == {"beta"}
+
+
+@pytest.mark.asyncio
+async def test_delete_history_set_handles_multiple_ids(
+    semantic_storage: SemanticStorage,
+    episode_storage,
+):
+    h1_id = await _add_episode(episode_storage, content="first")
+    h2_id = await _add_episode(episode_storage, content="second")
+    h3_id = await _add_episode(episode_storage, content="third")
+
+    associations = {
+        "alpha": [h1_id],
+        "beta": [h1_id, h2_id],
+        "gamma": [h3_id],
+        "delta": [h2_id],
+    }
+
+    for set_id, history_ids in associations.items():
+        for history_id in history_ids:
+            await semantic_storage.add_history_to_set(
+                set_id=set_id,
+                history_id=history_id,
+            )
+
+    assert await semantic_storage.get_history_messages_count(set_ids=None) == 5
+
+    await semantic_storage.delete_history_set(set_ids=["alpha", "beta", "missing"])
+
+    assert await semantic_storage.get_history_messages_count(set_ids=None) == 2
+    assert set(map(str, await semantic_storage.get_history_set_ids())) == {
+        "gamma",
+        "delta",
+    }
+    assert await semantic_storage.get_history_messages(set_ids=["gamma"]) == [h3_id]
+    assert await semantic_storage.get_history_messages(set_ids=["delta"]) == [h2_id]
+
+
+@pytest.mark.asyncio
 async def test_complex_feature_lifecycle(semantic_storage: SemanticStorage):
     embed = np.array([1.0] * 1536, dtype=float)
 
@@ -607,6 +666,50 @@ async def test_filter_by_metadata_nullity(semantic_storage: SemanticStorage):
         filter_expr=_expr("metadata.details IS NOT NULL"),
     )
     assert [feature.value for feature in not_null_results] == ["first"]
+
+
+@pytest.mark.asyncio
+async def test_get_feature_set_unknown_filter_column_errors(
+    semantic_storage: SemanticStorage,
+):
+    if getattr(semantic_storage, "backend_name", None) == "neo4j":
+        pytest.skip("Neo4j currently allows arbitrary property references")
+
+    await semantic_storage.add_feature(
+        set_id="user",
+        category_name="default",
+        feature="note",
+        value="first",
+        tag="misc",
+        embedding=np.array([1.0], dtype=float),
+    )
+
+    with pytest.raises(ValueError, match="Unsupported feature filter field"):
+        await semantic_storage.get_feature_set(
+            filter_expr=_expr("missing_column IN (foo)"),
+        )
+
+
+@pytest.mark.asyncio
+async def test_delete_feature_set_unknown_filter_column_errors(
+    semantic_storage: SemanticStorage,
+):
+    if getattr(semantic_storage, "backend_name", None) == "neo4j":
+        pytest.skip("Neo4j currently allows arbitrary property references")
+
+    await semantic_storage.add_feature(
+        set_id="user",
+        category_name="default",
+        feature="note",
+        value="first",
+        tag="misc",
+        embedding=np.array([1.0], dtype=float),
+    )
+
+    with pytest.raises(ValueError, match="Unsupported feature filter field"):
+        await semantic_storage.delete_feature_set(
+            filter_expr=_expr("missing_column IN (foo)"),
+        )
 
 
 @pytest.mark.asyncio

